File: VaquinhaAnimal.Api/VaquinhaAnimal.Domain/Entities/Campanha.cs
﻿using VaquinhaAnimal.Domain.Entities.Base;
using VaquinhaAnimal.Domain.Enums;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace VaquinhaAnimal.Domain.Entities
{
    public class Campanha : BaseEntity
    {
        [JsonPropertyName("data_criacao")]
        [Required]
        public DateTime DataCriacao { get; set; }                   

        [JsonPropertyName("data_inicio")]
        public DateTime? DataInicio { get; set; }

        [JsonPropertyName("duracao_dias")]
        public int DuracaoDias { get; set; }

        [JsonPropertyName("data_encerramento")]
        public DateTime? DataEncerramento { get; set; }

        [JsonPropertyName("titulo")]
        [Required]
        public string Titulo { get; set; }

        [JsonPropertyName("video_url")]
        public string VideoUrl { get; set; }

        [JsonPropertyName("descricao_curta")]
        [Required]
        public string DescricaoCurta { get; set; } // (200)

        [JsonPropertyName("descricao_longa")]
        [Required]
        public string DescricaoLonga { get; set; }// (MÍNIMO 500 MÁXIMO 5.000)

        [JsonPropertyName("valor_desejado")]
        [Required]
        public decimal ValorDesejado { get; set; } // MAIOR QUE ZERO

        [JsonPropertyName("total_arrecadado")]
        [Required]
        public decimal TotalArrecadado { get; set; } // MAIOR OU IGUAL A ZERO

        [JsonPropertyName("termos")]
        [Required]
        public bool Termos { get; set; }

        [JsonPropertyName("premium")]
        [Required]
        public bool Premium { get; set; }

        [JsonPropertyName("status_campanha")]
        [Required]
        public StatusCampanhaEnum StatusCampanha { get; set; }

        [JsonPropertyName("usuario_id")]
        [Required]
        public Guid Usuario_Id { get; set; }

        [JsonPropertyName("redes_sociais")]
        public List<RedeSocial> RedesSociais { get; set; }

        [JsonPropertyName("imagens")]
        public List<Imagem> Imagens { get; set; }

        [JsonPropertyName("conta_deposito")]
        public ContaDeposito ContaDeposito { get; set; }

        [JsonPropertyName("beneficiario")]
        public Beneficiario Beneficiario { get; set; }

        [JsonPropertyName("doacoes")]
        public List<Doacao> Doacoes { get; set; } 
    }
}

File: VaquinhaAnimal.App/src/app/campanhas/model/Campanha.ts
export class Campanha {
    id!: string;
    data_inicio?: Date;
    data_encerramento?: Date;
    duracao_dias!: number;
    titulo!: string;    
    descricao_curta!: string;
    descricao_longa!: string;
    valor_desejado!: number;
    total_arrecadado?: number;
    termos!: boolean;
    premium!: boolean;
    status_campanha?: number;
    percentual_arrecadado?: number;
    restam?: number;
    totalDoadores?: number;
    usuario_id!: string;
    video_url!: string;
    
    beneficiario!: Beneficiario;
    imagens!: Imagem[];
}

export class Beneficiario {
    id!: string;
    nome!: string;
    documento!: string;
    tipo!: string;
    codigo_banco!: string;
    numero_agencia!: string;
    digito_agencia!: string;
    numero_conta!: string;
    digito_conta!: string;
    tipo_conta!: string;
    recebedor_id!: string;
    campanha_id!: string;
}

export class Imagem {
    id!: string;
    tipo!: TipoImagemEnum;
    arquivo!: string;
    arquivo_upload!: string;
    campanha_id!: string;
}

export enum TipoImagemEnum {
    Capa = 1,
    Interna = 2
}